### PR TITLE
Issue#748 fixed:in binary mode only binary scores are shown

### PR DESCRIPTION
**Pull request by rimmakam**
Closes #748

### CODE REVIEW

**Review by kristinagalik — CHANGES_REQUESTED** [review]
hi @rimmakam :)

just a few things: 
1. what happens if you set the manual score of a pair to a negative value?
2. you could rather create a function that will be called on line 31 (for instance from the same place as `score_transplant_calculated` so you wouldn't need to deal with `ConfigParameters` in the `additive_scorer.py` file) 
3. please also add/fix tests so we can be sure it works :)

**Comment by rimmakam** [comment]
> hi @rimmakam :)
> 
> just a few things:
> 
> 1. what happens if you set the manual score of a pair to a negative value?
> 2. you could rather create a function that will be called on line 31 (for instance from the same place as `score_transplant_calculated` so you wouldn't need to deal with `ConfigParameters` in the `additive_scorer.py` file)
> 3. please also add/fix tests so we can be sure it works :)

Hello, @kristinagalik :)

I wanted to ensure if I understand that correctly: if a negative manual score is set for some pair, should that pair be excluded from the results or should it be kept but with binary score "-1" or "0"?

And also I wanted to ask if there are already some tests for the situation when we have manual scores with binary mode chosen (I haven't found it yet).

Thank you!

**Comment by kristinagalik** [comment]
@rimmakam 

if a negative value is set (or is negative by default) it means that a pair is not compatible. so by setting a negative manual value, you are forcing the pair to be incompatible (regardless of their initial level of compatibility), so it shouldn't display such a pair in the matchings.

so it essentially doesn't matter whether for incompatible manual scores you return a custom negative value or the inital or the set manual negative value, but for consistency i suggest -1 or something for everything negative in binary scoring? 

for this and maybe also for compatible pairs in binary scoring i would also suggest making new constants in scorer_constants.py. so 2 constants in total - negative and positive in binary matchings.

and for tests: i can't seem to find any, could you please add some tests :)?

**Review by kubantjan — CHANGES_REQUESTED** [review]
Waiting for the review be finished by kristina, then please rerequest my review :)

**Comment by rimmakam** [comment]
Hello, @kristinagalik,

I'm working on tests right now, but I'm a little bit lost - should I write them in test_hla_scorer or somewhere else? 

Thank you :)

**Comment by kristinagalik** [comment]
@rimmakam hmm maybe its better to write them in `test_configuration.py` :)

**Comment by rimmakam** [comment]
Hi, @kristinagalik. 

I would like to ask why the configuration parameters should be tested. It seemed to me that I should test whether the scoring works correctly given binary mode(on/off). For example when the manual score is negative and the binary mode is on, I would test if it results in overall score == "-1". Or I would test whether the score == "1" for a positive manual score given binary mode.

Thank you :)

**Comment by kristinagalik** [comment]
@rimmakam

so you could do it for instance in `test_ilp_solver.py`, that would be the best probably.

**Comment by rimmakam** [comment]
Hello, @kristinagalik,

I've written some tests, could you please take a look? :)

File: txmatching/scorers/hla_additive_scorer.py [hunk 1]
```diff
@@ -31,7 +32,15 @@ def max_transplant_score(self) -> float:
                     + self._configuration.blood_group_compatibility_bonus
         return max_value
 
-        # pylint: disable=too-many-return-statements
+    # in order to get the correct score in binary mode with manual score set, we should check few conditions
+    def get_score_when_manual_score_set(self, manual_score):
+        if self._configuration.use_binary_scoring:
+            if manual_score >= 0:
+                return POSITIVE_SCORE_BINARY_MODE
+            elif manual_score < 0:
```
Comment (kristinagalik) [inline txmatching/scorers/hla_additive_scorer.py:40]: maybe here is sufficient only `else`

File: tests/solvers/ilp_solver/test_ilp_solver.py [hunk 2]
```diff
@@ -230,6 +230,54 @@ def test_manual_scores_set_manual_score_of_every_pair_to_negative(self):
...
+        solution = solve_from_configuration(ConfigParameters(
+                                            solver_constructor_name=Solver.ILPSolver,
+                                            manual_donor_recipient_scores=[
+                                            ManualDonorRecipientScore(donor_db_id=3, recipient_db_id=2, score=10.0)],
+                                            use_binary_scoring=True
+            ), txm_event)
+
+        self.assertEqual(solution.calculated_matchings_list[0].score, 1.0)
+
+
```
Comment (kristinagalik) [inline tests/solvers/ilp_solver/test_ilp_solver.py:279]: redundant blank spaces between tests (functions in tests are for some reason only surrounded by 1 blank space)

Comment (kristinagalik) [reply tests/solvers/ilp_solver/test_ilp_solver.py:279]: also if you dont use a formatter, it would be good to use one :)

**Review by kristinagalik — CHANGES_REQUESTED** [review]
looks good, just some very small comments from me :) also please rebase. 

^ after this im approving :)

**Comment by rimmakam** [comment]
Hello, @kristinagalik,

changes have been added. I also wanted to ensure if I should rebase to master branch.

File: txmatching/scorers/hla_additive_scorer.py [hunk 3]
```diff
@@ -5,7 +5,8 @@
 from txmatching.configuration.subclasses import ForbiddenCountryCombination
 from txmatching.patients.patient import Donor, Recipient
 from txmatching.scorers.additive_scorer import AdditiveScorer
-from txmatching.scorers.scorer_constants import TRANSPLANT_IMPOSSIBLE_SCORE
+from txmatching.scorers.scorer_constants import TRANSPLANT_IMPOSSIBLE_SCORE, POSITIVE_SCORE_BINARY_MODE, \
+    NEGATIVE_SCORE_BINARY_MODE
```
Comment (kristinagalik) [inline txmatching/scorers/hla_additive_scorer.py:9]: sorry overlooked this, but imports should be alphabetically ordered, so is thould be `... import NEGATIVE_SCORE_BINARY_MODE, POSITIVE_SCORE_BINARY_MODE, TRANSPLANT_IMPOSSIBLE_SCORE`

**Review by kristinagalik — CHANGES_REQUESTED** [review]
hi @rimmakam, yes, please rebase to master. 

also sorry, one last change request. 

@kubantjan after this one change and rebase the code is fine by me, ready for your review :)

**Review by kristinagalik — APPROVED** [review]
@kubantjan za mna zmeny fajn :)

**Comment by kubantjan** [comment]
@kristinagalik use rerequest please :)

**Review by kubantjan — DISMISSED** [review]

**Comment by rimmakam** [comment]
Hello, @kristinagalik,

I'm facing some problems with rebase. I've never done it before and I'm not sure about how to do it correctly. I've tried just "git rebase master" and few other things, but it seems that nothing has changed. Could you please have a look if you see some changes? I would also appreciate any advice.

Thank you :)

**Comment by kristinagalik** [comment]
@rimmakam 
hmm after you did `git rebase master` did you also push? if yes, what IDE do you use? if you struggle further, maybe we can have a quick call?

**Comment by rimmakam** [comment]
Hello, @kristinagalik,

I didn't push, because I wasn't sure that I did the rebase correctly. I'm using PyCharm. I've tried to do it according to the official documentation, but unfortunately I'm still stuck, so if you had time to have a look, it would be great.

**Comment by kristinagalik** [comment]
Okay so there are multiple ways to do a rebase, 2 of those are: 
1. when you do a rebase through a console (on your branch just write `git rebase master`) you see stuff like 
```
First, rewinding head to replay your work on top of it...
Applying: Issue#748 fixed:in binary mode only binary scores are shown
Applying: Added function that deals with manual scores depending on whether the binary mode is set
Applying: Issue#748: tests added
Applying: Issue#748: last changes added
Applying: Issue#748: imports fixed
``` 
(If you didnt see anything like that, make sure to do to master branch, `git pull` and then checkout to your branch again and rebase.)
If you dont see any warnings, conflicts or anything, you can just do `git push -f` and the branch is rebased also remotely. 

2. when you are using pycharm, you can go to the master branch, at the bottom of the IDE go to "Git" 
image
and then right-click your branch, and pick the option to "checkout and rebase onto master". If you see no warnings and there was a pop up at the right bottom side of the screen saying something like "checked out and rebased onto master" you are good to `git push -f`

if you arent sure if you did the right thing and you are stuck between steps "rebase" and "push" just rebase again and it will tell you whether its up to date - if it is, you should be good to push.

if you worry that you did something along the way, i would just delete the branch, then create new one from the remote (`git switch Iss...`) and rebase again, but i dont think this should be needed

**Comment by rimmakam** [comment]
@kristinagalik, 

so I tried to do it by deleting the branch and creating new one, because nothing else worked, but I'm sure I did it the wrong way. I'm sorry :(

**Comment by kristinagalik** [comment]
oh sorry, it was meant to be deleted only locally, and then again restored from the remote. no worries,

**Comment by rimmakam** [comment]
Thank you very much, I hope it will be easier next time :) So we can leave it as it is?

**Review by kubantjan — APPROVED** [review]

**Pull request merged**